### ISSUE TITLE
Incompatible dictionary results in confusing error message

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#1183]
Dictionary literals with incompatible keys/values result in weird error message, because mypy translate the expression internally to something quite different before type checking.

```
from typing import Dict
d = {1 : ''} # type: Dict[int, int]
```

This gives a confusing error message:

```
t2.py:2: error: List item 0 has incompatible type "Tuple[int, str]"
```

**Comment by fabianhjr** [comment]
Ok, this has to do with `mypy/messages.py:466: msg = '{} item {} has incompatible type {}'.format(`; I will take a look at it. The only thing is that about 20 test cases will need to be changed.

**Comment by fabianhjr** [comment]
Ok, so this is what I have in mind:

Test if we are dealing with a List[Tuple[Any, Any]] and if so add `\n * NOTE: This might be a Dict` to the error message.

Additionally, also add the expected type.

**Comment by fabianhjr** [comment]
Ok, a fix is ready; It should now read:

`t2.py:2: error: List item 0 has incompatible type "Tuple[int, str]"; expected "Tuple[int, int]" * NOTE: This might be a Dict[K, V] instead of List[Tuple[K, V]]`

**Comment by JukkaL** [comment]
Fixed by  #3100.
